File: Solutions/Python/clone_graph.py
"""
    [MEDIUM]
    133. Clone Graph

    Concepts:
    - graph
    - breadth-first search

    Stats:
        Runtime | 42 ms     [Beats 55.70%]
        Memory  | 18.40 MB  [Beats 7.44%]
"""

"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
from collections import defaultdict, deque

class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if node is None or not node.neighbors:
            return None if node is None else Node(val=node.val)

        adj = defaultdict(set)  # int |-> set[int]

        # explore graph and update adjacency list
        queue = deque([node])
        while queue:
            len_q = len(queue)

            for _ in range(len_q):
                curr = queue.popleft()

                for n in curr.neighbors:
                    # if neighbor not explored...
                    if not adj[n]:
                        queue.append(n)
                        
                    adj[curr].add(n)

        # rebuild graph from adjacency list
        nodes = {n.val : Node(val=n.val) for n in adj}

        for nde, nei in adj.items():
            for nbr in nei:
                nodes[nde.val].neighbors.append(nodes[nbr.val])

        return nodes[node.val]

File: Solutions/Cpp/clone_graph.cpp
/*
    [MEDIUM]
    133. Clone Graph

    Concepts:
    - graph
    - breadth-first search

    Stats:
        Runtime | 3 ms      [Beats 67.71%]
        Memory  | 12.15 MB  [Beats 47.01%]
*/

/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (!node || node->neighbors.empty())
            return (node) ? (new Node(node->val)) : (nullptr);

        std::unordered_map<Node*, std::unordered_set<Node*>> adj = {};

        // explore graph and update adjacency list
        std::queue<Node*> q;
        q.push(node);
        int len_q;
        Node* curr;

        while (!q.empty()) {
            len_q = q.size();

            for (int _ = 0; _ < len_q; _++) {
                curr = q.front();
                q.pop();

                for (Node* n : curr->neighbors) {
                    // if neighbor not explored...
                    if (adj[n].empty())
                        q.push(n);

                    adj[curr].insert(n);
                }
            }
        }

        // rebuild graph from adjacency list
        std::unordered_map<int, Node*> nodes = {};
        for (auto& [n, _] : adj)
            nodes[n->val] = new Node(n->val);

        for (auto& [nde, nei] : adj) {
            for (Node* nbr : nei)
                nodes[nde->val]->neighbors.push_back(nodes[nbr->val]);
        }

        return nodes[node->val];
    }
};